test/fixtures/basic-types/types.ts

```diff
@@ -309,6 +309,22 @@ describe('composables', () => {
     expectTypeOf(useFetch('/test', { default: () => 500 }).data).toEqualTypeOf<Ref<unknown>>()
   })
 
+  it('correct types when using ResT type-assertion with default function', () => {
+    // @ts-expect-error default type should match generic type
+    useFetch<string>('/test', { default: () => 0 })
+    // @ts-expect-error default type should match generic type
+    useLazyFetch<string>('/test', { default: () => 0 })
+    // @ts-expect-error default type should match generic type
+    useAsyncData<string>(() => $fetch('/test'), { default: () => 0 })
+    // @ts-expect-error default type should match generic type
+    useLazyAsyncData<string>(() => $fetch('/test'), { default: () => 0 })
+
+    expectTypeOf(useFetch<string>('/test', { default: () => 'test' }).data).toEqualTypeOf<Ref<string>>()
+    expectTypeOf(useLazyFetch<string>('/test', { default: () => 'test' }).data).toEqualTypeOf<Ref<string>>()
+    expectTypeOf(useAsyncData<string>(() => $fetch('/test'), { default: () => 'test' }).data).toEqualTypeOf<Ref<string>>()
+    expectTypeOf(useLazyAsyncData<string>(() => $fetch('/test'), { default: () => 'test' }).data).toEqualTypeOf<Ref<string>>()
```

Comment:
nice. could you also add tests with both default + transform so we verify those still work?

Reply:
I just noticed that providing a generic parameter results in the `transform` not being able to return a different type.
e. g. this is not possible because `number` isn't assignable to `string`
```ts
useFetch<string>('/test', {default: () => 'test', transform: (arg) => 0})
```
Unfortunately this seems to be a typescript limitation at the moment, see https://github.com/microsoft/TypeScript/issues/14400

---
Should I simply add tests for the case of returning the matching type?

Reply:
As long as it's not a regression in this PR, I'm happy just documenting current behaviour in the tests 👍

Reply:
Added some more tests and there shouldn't be any regressions :)